### PR TITLE
Add scale subresource status to the OpenTelemetryCollector CRD status

### LINKED ISSUE #775: Add missing scale subresource status in order to use an HPA resource over an OpenTelemetryCollector CR
The `OpenTelemetryCollector` has a configuration field named `replicas` (see [openTelemetryCollector.spec.replicas](https://github.com/open-telemetry/opentelemetry-operator/blob/main/apis/v1alpha1/opentelemetrycollector_types.go#L38)). However it lacks some status info in order to configure an HPA on a it (see [scale subresource](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#scale-subresource) documentation) using the defined `scale` subresource in the `OpenTelemetryCollector` CRD.

### Suggestion

I suggest to update the `OpenTelemetryCollector` CRD status with:

```golang
// ScaleSubresourceStatus defines the observed state of the OpenTelemetryCollector's
// scale subresource.
type ScaleSubresourceStatus struct {
	// The total number of non-terminated pods targeted by this
	// OpenTelemetryCollector's deployment or statefulSet.
	// +optional
	Replicas int32 `json:"replicas,omitempty"`

	// The selector used to match the OpenTelemetryCollector's
	// deployment or statefulSet pods.
	// +optional
	Selector string `json:"selector,omitempty"`
}

// OpenTelemetryCollectorStatus defines the observed state of OpenTelemetryCollector.
type OpenTelemetryCollectorStatus struct {
	[...]

	// Scale is the OpenTelemetryCollector's scale subresource status.
	// +optional
	Scale ScaleSubresourceStatus `json:"scale,omitempty"`

	[...]
}

// +kubebuilder:object:root=true
// +kubebuilder:resource:shortName=otelcol;otelcols
// +kubebuilder:subresource:status 
// +kubebuilder:subresource:scale:specpath=.spec.replicas,statuspath=.status.scale.replicas,selectorpath=.status.scale.selector
[...]
```

> **NOTE:** The work that was done in PR #746 doesn't fulfill our needs. We need to scale based on the memory consumption. We may also in the future configure scaling based on a custom metric. We also want to configure our own desired [scaling behavior](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#configurable-scaling-behavior).

### PR DESCRIPTION
**Pull request by secat**
Resolves #775

### CODE REVIEW

File: apis/v1alpha1/opentelemetrycollector_types.go [hunk 1]
```diff
@@ -129,12 +129,30 @@ type OpenTelemetryTargetAllocator struct {
...
+	// deployment or statefulSet pods.
+	// +optional
+	Selector string `json:"selector,omitempty"`
+}
+
 // OpenTelemetryCollectorStatus defines the observed state of OpenTelemetryCollector.
 type OpenTelemetryCollectorStatus struct {
 	// Replicas is currently not being set and might be removed in the next version.
 	// +optional
 	Replicas int32 `json:"replicas,omitempty"`
```
[the remark this answers is not in the document]
Comment (secat) [reply apis/v1alpha1/opentelemetrycollector_types.go:150]: Exactly, should I remove it in this PR?

[the remark this answers is not in the document]
Comment (secat) [reply apis/v1alpha1/opentelemetrycollector_types.go:150]: This status field is currently unused

[the remark this answers is not in the document]
Comment (jpkrohling) [reply apis/v1alpha1/opentelemetrycollector_types.go:150]: Right, but I think it's still best to deprecate in one version, removing in another. Even if only to serve as a pattern for future, similar tasks.

**Review by jpkrohling — APPROVED** [review]

**Comment by secat** [comment]
Fixed also a `golang-ci` lint error: disabled the `exhaustive` linter over the `switch mode` located in the `updateScaleSubResourceStatus` function

**Comment by pavolloffay** [comment]
@secat please take a look here https://github.com/open-telemetry/opentelemetry-operator/issues/775#issuecomment-1074249189

**Comment by secat** [comment]
@pavolloffay and @jpkrohling is there anything that needs to be done prior to merge this PR? Thank you in advance! 😃

**Comment by jpkrohling** [comment]
This is good to go for me, but given @pavolloffay's comments, it would be good to have an explicit approval from him.

**Comment by secat** [comment]
@pavolloffay anything missing in this PR? Thank you in advance

**Comment by pavolloffay** [comment]
@secat sorry for the dealy, first I was late on the review due to playing with the HPA and then I got sick.

I think we can merge this but let's discuss one more thing about HPA in #775

**Review by pavolloffay — APPROVED** [review]

File: apis/v1alpha1/opentelemetrycollector_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pavolloffay) [reply apis/v1alpha1/opentelemetrycollector_types.go:151]: As far as I know, properties from the CR should not be removed. The cleanup can be done only when we promote the CRD to a new version.

**Comment by pavolloffay** [comment]
I think this can go in as it is. I would maybe prefer to have the scale fields directly in the status to avoid deprecating  `status.replicas` and inconsistency between `status.scale.replicas`.

**Pull request merged**